### DISCUSSION TITLE
How to move a value out of tokio RwLock/Mutex

### DISCUSSION THREAD

**Discussion by ActivePeter** [tokio-rs/tokio#4047] [Q&A]
I'm using protobuf and it need the vec to construct bytes, but the vec is in RwLock and cannot be moved. Is there any way to move the data without cloning the vec?

**Comment by hawkw** [comment]
Try using [`std::mem::take`](https://doc.rust-lang.org/stable/std/mem/fn.take.html);
```rust
use tokio::sync::RwLock;
use std::mem;

let rwlock = RwLock::new(Vec::new());

// ...

// use `mem::take` to move the `Vec` out of the `RwLock`.
let mut write = lock.write().await;
let the_vec = mem::take(&mut *write);

// now, we have the `Vec` and can pass it into whatever function we need
do_something_with_the_vec(the_vec);
```

`take` will move the `Vec` out of the `RwLock`, replacing it with the value of `Vec::default` (an empty vector), and return it. The `RwLock` will then contain a new empty `Vec`, which you can use again. 

Alternatively, if you want to replace the value inside the `RwLock` with something other than `Vec::default`, you can use [`std::mem::replace`](https://doc.rust-lang.org/stable/std/mem/fn.replace.html) instead. This functions similarly to `mem::take`, but it allows you to provide a replacement value.

**Reply by ActivePeter** [reply]
Thank you so much.

**Comment by Darksonn** [comment]
For cases where you have ownership of the lock, you can call the `into_inner` method which consumes the lock and returns the contents.
